src/transaction/atomic_replication_helper.hpp

```diff
@@ -96,6 +99,49 @@ namespace cublog
       std::map<TRANID, vpid_set_type> m_atomic_sequences_vpids_map;
 #endif
   };
+
+  template <typename T>
+  int atomic_replication_helper::add_atomic_replication_unit (THREAD_ENTRY *thread_p, TRANID tranid, log_lsa record_lsa,
+      LOG_RCVINDEX rcvindex, VPID vpid, log_rv_redo_context &redo_context, const log_rv_redo_rec_info<T> &record_info)
+  {
+#if !defined (NDEBUG)
+    if (!VPID_ISNULL (&vpid) && !check_for_page_validity (vpid, tranid))
+      {
+	// the page is no longer relevant
+	assert (false);
+      }
+    vpid_set_type &vpids = m_atomic_sequences_vpids_map[tranid];
+    vpids.insert (vpid);
+#endif
+
+    m_atomic_sequences_map[tranid].emplace_back (atomic_replication_unit (record_lsa, vpid, rcvindex));
+    int error_code = m_atomic_sequences_map[tranid].back ().fix_page (thread_p);
+    if (error_code != NO_ERROR)
+      {
+	return error_code;
+      }
+
+    m_atomic_sequences_map[tranid].back ().apply_log_redo (thread_p, redo_context, record_info);
```

Comment:
This function is called "add_atomic_replication_unit", but here we also "apply" the replication unit.
My understanding is that the atomic sequence is composed of atomic unit (a kind of wrapper on top of log record), the entire atomic sequence needs to be applied at once. For this, I thought we need to fix all pages first, then apply the records, then unfix all pages.

This code seems to apply the logs as the pages are fixed; this comes with a possibility that a page fix (with ordered fix) may trigger an unfix in a previously fixed page from the sequence, allowing other transaction to change it.